src/main/kotlin/org/elm/workspace/ElmFormatCLI.kt

```diff
@@ -1,27 +1,82 @@
 package org.elm.workspace
 
+import com.intellij.execution.ExecutionException
 import com.intellij.execution.process.ProcessOutput
-import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiFile
+import org.elm.lang.core.psi.ElmFile
 import org.elm.openapiext.GeneralCommandLine
 import org.elm.openapiext.execute
-import org.elm.openapiext.pathAsPath
-import org.elm.openapiext.withWorkDirectory
+import org.elm.openapiext.isSuccess
+import org.elm.openapiext.toPsiFile
 import java.nio.file.Path
 
 class ElmFormatCLI(val elmFormatExecutablePath: Path) {
 
-    fun reformatFile(project: Project, elmVersion: Version, file: VirtualFile, owner: Disposable = project): ProcessOutput {
+    fun reformatDocument(elmVersion: Version, document: Document): ProcessOutput {
         val arguments = listOf(
                 "--yes",
-                "--elm-version",
-                "${elmVersion.x}.${elmVersion.y}",
-                file.path
+                "--elm-version=${elmVersion.x}.${elmVersion.y}",
+                "--stdin"
         )
+
         return GeneralCommandLine(elmFormatExecutablePath)
-                .withWorkDirectory(file.parent.pathAsPath)
                 .withParameters(arguments)
-                .execute(owner)
+                .execute(document.text)
     }
+
+    companion object {
+
+        fun getElmVersion(project: Project, file: VirtualFile): Version? {
+            val psiFile = file.toPsiFile(project) ?: return null
+            return getElmVersion(psiFile)
+        }
+
+        fun getElmVersion(psiFile: PsiFile): Version? {
+            if (psiFile !is ElmFile) return null
+            val elmVersion: Version = when (val elmProject = psiFile.elmProject) {
+                is ElmApplicationProject -> elmProject.elmVersion
+                is ElmPackageProject -> elmProject.elmVersion.low
+                else -> return null
+            }
+
+            return elmVersion
+        }
+
+        fun formatDocumentAndSetText(ctx: Context) {
+
+            val file = FileDocumentManager.getInstance().getFile(ctx.document)
```

Comment:
What's the point of getting the VirtualFile here? Isn't it in the context object?